### ISSUE TITLE
Super issue

### ISSUE THREAD

**Issue by Apollo3zehn** [Apollo3zehn/ManagedEtherCAT#1]
- [ ] Scan devices or manual list: then code generation, strongly typed code (pdos). If user then uses allocation-free code real-time capabilities are possible.

**Comment by ToGoOrNotToGo** [comment]
Hi. Is there still a plan for a pure C# EtherCAT implementation? If so, I recommend using System.Threading.Tasks.Dataflow and Microsoft.IO.RecyclableMemoryStream.

**Comment by Apollo3zehn** [comment]
Yes I do plan it, but first I would like to bring my current project (HDF5.NET) to a beta state (soon). However, I do not give any promises about the timeline as this just a hobby project.

My dream is to have a pure C# implementation which is divided into a configuration part with no restrictions and a separate real-time C# process where no allocation is allowed to not trigger the GC. In combination with the real-time Linux kernel patch this should be able to compete with C implementations.

Since I just worked with C# source generators in my current project, maybe they can be applied here, too. So that a strongly typed EtherCAT network type hierarchy is source generated by discovering the connected slaves ... just ideas.

I have much worked with TwinCAT and SOEM and they work fine of course, but debugging is a pain. I hope this project will make everything much easier.

Thanks for you suggestions, I will consider them when I start with the project.

**Comment by ToGoOrNotToGo** [comment]
That would be very interesting! Do you have more information about using real-time Linux kernel (do you mean WSL2?) in an application?

**Comment by Apollo3zehn** [comment]
I am talking about this project: https://wiki.linuxfoundation.org/realtime/start

I have used it a few years ago on a notebook and got very nice results. At that time I had to compile the kernel by myself (which is not possible on WSL2 because of some NTFS issues). But on a real Linux this is easy. I think enabling realtime capabilities is a simple kernel config option now. A few years ago that was a patch, not a single config option.

**Comment by ToGoOrNotToGo** [comment]
There is now another way to enable a process or thread to run as a real-time kernel in Windows:
https://learn.microsoft.com/en-us/windows/iot/iot-enterprise/soft-real-time/soft-real-time-device#use-mdm-bridge-wmi-provider-to-configure-the-windowsiot-csp
https://github.com/amitxv/ReservedCpuSets

Do you have experience with that?

**Comment by Apollo3zehn** [comment]
Thank for the links, very interesting read. I have no experience with this new real-time configuration options.

When I find the time to start the project I will begin with Linux as I find it more suitable for such kind of tasks. When that works I think it is not too difficult to make it also run on Windows IOT.
